gbpservice/neutron/nsf/proxy/proxy.py

```diff
@@ -0,0 +1,223 @@
+        self.socket.listen(self.max_connections)
+
+    def listen(self):
+        client, address = self.socket.accept()
+        self.proxy.new_client(client, address)
+
+
+class TcpClient(object):
+
+    def __init__(self, conf, proxy):
+        self.conf = conf
+        self.proxy = proxy
+        self.server_address = conf.rest_server_address
+        self.server_port = conf.rest_server_port
+        # Connect the socket to the port where the server is listening
+        self.server = (self.server_address, self.server_port)
+
+    def connect(self):
+        sock = socket.socket()
+        print >>sys.stderr, 'connecting to %s port %s' % self.server
+        sock.settimeout(self.conf.connect_max_wait_timeout)
+        try:
+            sock.connect(self.server)
+        except socket.error, exc:
+            print "Caught exception socket.error : %s" % exc
+            return sock, False
+        return sock, True
+
+
+class Connection(object):
+
+    def __init__(self, conf, socket):
+        self._socket = socket
+        self._idle_wait = conf.idle_min_wait_timeout
+        self._idle_timeout = conf.idle_max_wait_timeout
+        self._idle_count_max = (self._idle_timeout / self._idle_wait)
+        self._idle_count = 0
+
+    def _tick(self):
+        self._idle_count += 1
+
+    def _timedout(self):
+        if self._idle_count > self._idle_count_max:
+            raise ConnectionTimedOut
+
+    def idle(self):
+        self._tick()
+        self._timedout()
+
+    def idle_reset(self):
+        self._idle_count = 0
+
+    def recv(self):
+        self._socket.settimeout(self._idle_wait)
+        try:
+            data = self._socket.recv(16)
+            if data and len(data):
+                self.idle_reset()
+                return data
+            self.idle()
+        except socket.timeout:
+            self.idle()
+        return None
+
+    def send(self, data):
+        self._socket.send(data)
+
+    def close(self):
+        self._socket.close()
+
+    def identify(self):
+        return self._socket.fileno()
+
+
+class ProxyConnection(object):
+
+    def __init__(self, conf, unix_socket, tcp_socket):
+        self._unix_conn = Connection(conf, unix_socket)
+        self._tcp_conn = Connection(conf, tcp_socket)
+
+    def close(self):
+        self._unix_conn.close()
+        self._tcp_conn.close()
+
+    def _proxy(self, rxconn, txconn):
+        data = rxconn.recv()
+        if data:
+            txconn.send(data)
+
+    def run(self):
+        try:
+            self._proxy(self._unix_conn, self._tcp_conn)
+            self._proxy(self._tcp_conn, self._unix_conn)
+            return True
+        except:
+            self._unix_conn.close()
+            self._tcp_conn.close()
+            return False
+
+    def identify(self):
+        return '%d:%d' % (
+            self._unix_conn.identify(),
+            self._tcp_conn.identify())
+
+
+class Worker(object):
+
+    def run(self):
+        while True:
```

Comment:
No it will not, as we are scheduling out this thread everytime using time.sleep(0)